File: src/main/java/faba/java/source.java
package faba.java;


import org.objectweb.asm.ClassReader;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.Enumeration;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;

interface Source {
    void process(Processor processor);
}

final class JarFileSource implements Source {
    final File file;

    JarFileSource(File file) {
        this.file = file;
    }

    @Override
    public void process(Processor processor) {
        try {
            JarFile jarFile = new JarFile(file);
            Enumeration<JarEntry> entries = jarFile.entries();
            while (entries.hasMoreElements()) {
                JarEntry entry = entries.nextElement();
                if (entry.getName().endsWith(".class")) {
                    try (InputStream is = jarFile.getInputStream(entry)) {
                        processor.processClass(new ClassReader(is));
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

interface Processor {
    void processClass(ClassReader classReader);
}

File: src/main/scala/source.scala
package faba.source

import java.io.{File, FileInputStream}
import java.util.jar.JarFile

import org.objectweb.asm._

import scala.collection.JavaConverters._
import scala.language.existentials

sealed trait Source {
  def process(processor: Processor): Unit
}

case class ClassSource(classes: Class[_]*) extends Source {
  override def process(processor: Processor): Unit =
    classes.foreach { clazz =>
      processor.processClass(new ClassReader(clazz.getCanonicalName))
    }
}

case class FileSource(file: File) extends Source {
  override def process(processor: Processor): Unit = {
    val is = new FileInputStream(file)
    try {
      processor.processClass(new ClassReader(is))
    } finally {
      is.close()
    }
  }
  override def toString = file.toString
}

case class JarFileSource(file: File) extends Source {
  override def process(processor: Processor): Unit = {
    val jarFile = new JarFile(file)
    for (entry <- jarFile.entries().asScala) {
      if (entry.getName.endsWith(".class")) {
        val is = jarFile.getInputStream(entry)
        try {
          processor.processClass(new ClassReader(is))
        } finally {
          is.close()
        }
      }
    }
  }

  override def toString = file.toString
}

case class MixedSource(sources: List[Source]) extends Source {
  override def process(processor: Processor): Unit =
    sources.foreach { s =>
      println(s"${new java.util.Date} processing $s")
      s.process(processor)
    }
}

trait Processor {
  def processClass(classReader: ClassReader): Unit
}
